data/external/scrapers/public_transport.py

```diff
@@ -0,0 +1,66 @@
+from pathlib import Path
+from external.scraping_utils import cached_json
+from external.scrapers.roomfinder import scrape_maps
+import csv
+from decimal import Decimal
+from geopy.distance import distance
+
+# CSV indexes
+STATIONID = 0
+NAME = 1
+ORT = 2
+GLOBAL_ID = 3
+WGS84X = 4  # lat
+WSG84Y = 6  # lon
+
+MAXDISTANCE = 1  # max distance from building
+
+
+def avg(x1, x2):
+    return (x1 + x2) / 2
+
+
+def nearby(building_coords:tuple, stations: list[dict]) -> list[dict]:
+    results=[]
+    for station in stations:
+        distance_to_building=round(distance((station.get("lat"), station.get("lon")), (building_coords)).km,2)
+        if distance_to_building <= MAXDISTANCE:
+            station["lat"]=str(station.get("lat"))  #decimal to string to allow json serialization
+            station["lon"]=str(station.get("lon"))
+            station["distance"]=distance_to_building
+            results.append(station)
+    return results
+
+
+@cached_json("public_transport.json")
+def scrape_stations():
+    with Path("scrapers/MVV_HSTReport2212.csv").open("r") as file:
+        lines = csv.reader(file, delimiter=";")  
+        next(lines) # skip first line as it contains row names
+        lines = filter(lambda l: not all(not bool(i) for i in l), lines)  # filter out lines where each value is ''
+        stations = list(
+            map(
+                lambda station: {
+                    "id":station[STATIONID],
+                    "name":station[NAME],
+                    "ort":station[ORT],
+                    "global-id":station[GLOBAL_ID],
+                    "lat":Decimal(station[WGS84X].replace(",", ".")),
+                    "lon":Decimal(station[WSG84Y].replace(",", ".")),
+                },
+                lines,
+            )
+        )
+    buildings = scrape_maps()
+    building_coords: list = []
+    for building in buildings:
+        # ignore entries that dont have assigned lon/lat
+        if latlonbox := building.get("latlonbox"):
+            # since lat and lon have rather long decimal points it is probably best to not use floats
+            latn, lats = Decimal(latlonbox.get("north")), Decimal(latlonbox.get("south"))
+            lone, lonw = Decimal(latlonbox["east"]), Decimal(latlonbox["west"])
+            lat: Decimal = avg(latn, lats)
+            lon: Decimal = avg(lone, lonw)
+            building_coords.append({"id": building.get("id"), "desc": building.get("desc"), "center-coord": {"lat":str(lat),"lon":str(lon)}, "public-transport":nearby((lat,lon),stations)})
+
```

Comment:
Conceptually, the scrapers should only scrape the data, which is then merged into the data pipeline.
If necessary, a scraper can rely upon itself, but not other parts.

The benefit of this is that the scrapers are more decoupled and hopefully easier to maintain.

```suggestion
import csv
from decimal import Decimal

from external.scraping_utils import cached_json


@cached_json("public_transport.json")
def scrape_stations():
    with open("scrapers/MVV_HSTReport2212.csv", "r", newline='') as file:
        reader = csv.DictReader(file, delimiter=";")
        return [{
                    "id": int(line["﻿HstNummer"]),
                    "name":line["Name ohne Ort"],
                    "location":line["Ort"],
                    "global_id":line["Globale ID"],
                    "lat":Decimal(line["WGS84 X"].replace(",", ".")),
                    "lon":Decimal(line["WGS84 Y"].replace(",", ".")),
                } for line in reader if line["﻿HstNummer"]]

```

Reply:
your proposed change would just change the data format from csv to json though? looking through all the stations and filtering them by distance does take some time, so I think it would be quite beneficial to have the results cached on the server. I do understand your concern about decoupling, however I feel like, in this scenario it is almost mandatory.

Reply:
> looking through all the stations and filtering them by distance does take some time

That is fine. The main pipeline takes about 2m to complete. Most of this time is spend in image resizing.

Why I am not really concerned with the extra 78.4s is, that I don't think that the current algorithm (looping over every coordinate for every entry is n*m, which is pretty bad)

I think the standard techniques can help us a lot and should be able to get this into the regime where it is neglable:
- Your current model of the world is not necessary: using geodesic is accurate, but for small distances, assuming the earth is flat is fine.
  For example, the following runs in acceptable time.
  ```py
  def nearby(building_coords:tuple, stations: list[dict]) -> list[dict]:
    results=[]
    for station in stations:
        station_coords=(station.get("lat"), station.get("lon"))
        distance_to_building=great_circle(station_coords, building_coords).km
        if distance_to_building <= MAXDISTANCE:
            station["lat"]=str(station.get("lat"))  #decimal to string to allow json serialization
            station["lon"]=str(station.get("lon"))
            station["distance"]=geodesic(station_coords, building_coords).m
            results.append(station)
    return results
    ```
- bucketing/sorting: We really don't need to compare every item against every other station, finding a range via binary search should be enough.
- Caching: in development images are only resized when nessesary, this could be a factor here to, but I doubt that it is nessesary.

>  I do understand your concern about decoupling, however I feel like, in this scenario it is almost mandatory.

Another argument why separating them is a good idea is that we at this point (before running the pipeline) don't have a good overview of what buildings exist.

Reply:
I'm sorry, but I just dont understand at what point you would want to run this, then. If your suggestion is in python, then it must be part of the pipeline, as that is the only part written in python. Or am I misunderstanding something here?

Reply:
Yes, the code snippet is a proposed alternative for the `nearby` function from your code.

Reply:
I got that, but what confuses me is that this proposed change:
```
@cached_json("public_transport.json")
def scrape_stations():
    with open("scrapers/MVV_HSTReport2212.csv", "r", newline='') as file:
        reader = csv.DictReader(file, delimiter=";")
        return [{
                    "id": int(line["﻿HstNummer"]),
                    "name":line["Name ohne Ort"],
                    "location":line["Ort"],
                    "global_id":line["Globale ID"],
                    "lat":Decimal(line["WGS84 X"].replace(",", ".")),
                    "lon":Decimal(line["WGS84 Y"].replace(",", ".")),
                } for line in reader if line["﻿HstNummer"]]
```
does not use that function at all. Should I make another json file that links building to station?

Reply:
No, I don't think the matching needs to be statically cached.
I would like the building-matching code to live in `processor/public_transport.py`, as this has less coupling.

Reply:
I have now changed to using the other csv as only that one includes the stationID required for the timeplan js-lib. 
This one includes substations as well, so I put them in a list with in the parent station. I chose to return it as a dict for now, so it is easier to look up substations. let me know if you prefer a list though. 
Also, I added typing and a test.